File: sdks/typescript/src/page_info_values.ts
import type { ToolResultData } from "./types.js";
import { ToolResultValue } from "./helpers.js";

/** Pagination metadata projection without owning collection state. */
export class PageInfo {
  public constructor(
    public readonly nextCursor: string | null = null,
    public readonly truncated = false,
    public readonly totalKnown = false,
    public readonly total: bigint | null = null,
    public readonly returned: bigint = 0n,
  ) {}

  public applyTo(result: ToolResultValue): ToolResultValue {
    let updated = result
      .withTruncated(result.truncated || this.truncated)
      .withMeta("page.truncated", String(this.truncated))
      .withMeta("page.total_known", String(this.totalKnown))
      .withMeta("page.returned", this.returned.toString());
    if (this.total !== null) updated = updated.withMeta("page.total", this.total.toString());
    if (this.nextCursor !== null) {
      updated = updated.withMeta("page.next_cursor", this.nextCursor);
      const continuation = JSON.stringify({
        next_cursor: this.nextCursor,
        returned: this.returned.toString(),
        total: this.total === null ? null : this.total.toString(),
        total_known: this.totalKnown,
        truncated: true,
      });
      updated = updated.withOutput(`${updated.output}\n[page]${continuation}`);
    }
    return updated;
  }
}

File: sdks/python/src/echo_agent_sdk/page_info_values.py
from __future__ import annotations

import json
from dataclasses import dataclass

from .intrinsics import ToolResult


@dataclass(frozen=True, slots=True)
class PageInfo:
    next_cursor: str | None = None
    truncated: bool = False
    total_known: bool = False
    total: int | None = None
    returned: int = 0

    def apply_to(self, result: ToolResult) -> ToolResult:
        updated = result.with_truncated(result.truncated or self.truncated)
        updated = updated.with_meta("page.truncated", str(self.truncated).lower())
        updated = updated.with_meta("page.total_known", str(self.total_known).lower())
        updated = updated.with_meta("page.returned", str(self.returned))
        if self.total is not None:
            updated = updated.with_meta("page.total", str(self.total))
        if self.next_cursor is not None:
            updated = updated.with_meta("page.next_cursor", self.next_cursor)
            continuation = json.dumps(
                {
                    "next_cursor": self.next_cursor,
                    "returned": self.returned,
                    "total": self.total,
                    "total_known": self.total_known,
                    "truncated": True,
                },
                separators=(",", ":"),
            )
            updated = updated.with_output(f"{updated.output}\n[page]{continuation}")
        return updated
